File: appunvs/ios/Runtime/State/AppState.swift
// AppState — host-wide observable state.  Currently carries the user's
// theme override (light / dark / nil = follow system).  Persists to
// UserDefaults so the choice survives restarts.
//
// Extends naturally to hold the active Box reference, the relay
// connection status, and the auth tokens once the network slice lands.
import Foundation
import SwiftUI
import Combine

@MainActor
final class AppState: ObservableObject {

    enum ThemeOverride: String, CaseIterable, Identifiable {
        case system, light, dark
        var id: String { rawValue }
    }

    /// User-selected theme override.  `system` defers to OS Dynamic Type
    /// trait; `light` / `dark` force the corresponding scheme.
    @Published var themeOverride: ThemeOverride {
        didSet { Self.persistTheme(themeOverride) }
    }

    /// Default model id used by the chat composer when the user hasn't
    /// picked a per-turn override.  Stored as the raw provider id (e.g.
    /// "deepseek-chat"); ModelCatalog.find resolves it to a display
    /// entry.  Empty string means "use the relay's engine default".
    @Published var defaultModelID: String {
        didSet { Self.persistModel(defaultModelID) }
    }

    /// Resolved value handed to `.preferredColorScheme(...)` on the root
    /// view.  `nil` means "let the system decide".
    var preferredColorScheme: ColorScheme? {
        switch themeOverride {
        case .system: return nil
        case .light:  return .light
        case .dark:   return .dark
        }
    }

    init() {
        themeOverride = Self.loadTheme()
        defaultModelID = Self.loadModel()
    }

    // MARK: - persistence

    private static let themeKey = "appunvs.theme.override"
    private static let modelKey = "appunvs.model.default"

    private static func loadTheme() -> ThemeOverride {
        guard let raw = UserDefaults.standard.string(forKey: themeKey),
              let parsed = ThemeOverride(rawValue: raw)
        else { return .system }
        return parsed
    }

    private static func persistTheme(_ value: ThemeOverride) {
        UserDefaults.standard.set(value.rawValue, forKey: themeKey)
    }

    private static func loadModel() -> String {
        // Validate against the catalog so a release that drops an old id
        // doesn't leave the user pinned to a model the relay no longer
        // accepts; fall back to the first catalog entry instead.
        let raw = UserDefaults.standard.string(forKey: modelKey) ?? ""
        if raw.isEmpty { return ModelCatalog.fallback.id }
        if ModelCatalog.find(id: raw) != nil { return raw }
        return ModelCatalog.fallback.id
    }

    private static func persistModel(_ value: String) {
        UserDefaults.standard.set(value, forKey: modelKey)
    }
}

File: appunvs/android/app/src/main/java/com/appunvs/runtime/state/AppState.kt
// AppState — host-wide observable state held in a ViewModel.
// Currently carries the theme override (system / light / dark).
// DataStore Preferences provides a small async key-value store; we
// load on init and persist on every override change.
//
// Extends naturally to active Box reference, auth tokens, relay
// connection status as those slices land.
package com.appunvs.runtime.state

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.launch

/// Module-level DataStore handle.  Names the on-disk file that backs
/// our Preferences; keep distinct from any other DataStore the app
/// might add later.
private val Application.dataStore by preferencesDataStore(name = "appunvs.prefs")

/// AndroidViewModel because we need an Application reference for the
/// DataStore handle.  Compose calls `viewModel()` and gets the same
/// instance for the lifetime of the activity.
class AppState(application: Application) : AndroidViewModel(application) {

    enum class ThemeOverride { SYSTEM, LIGHT, DARK }

    private val themeKey = stringPreferencesKey("theme.override")
    private val modelKey = stringPreferencesKey("model.default")
    private val store: androidx.datastore.core.DataStore<Preferences> = application.dataStore

    /// Compose-observable mirror of the persisted override.  Updated on
    /// init from disk, and on every set call (which also persists).
    var themeOverride by mutableStateOf(ThemeOverride.SYSTEM)
        private set

    /// Default chat model id (e.g. "deepseek-chat") — what the composer
    /// uses when the user hasn't set a per-turn override on the chat
    /// header.  Persisted via DataStore.  Unknown ids fall back to
    /// ModelCatalog.fallback.id so a release that drops a model doesn't
    /// strand existing users.
    var defaultModelID by mutableStateOf(ModelCatalog.fallback.id)
        private set

    init {
        viewModelScope.launch {
            val prefs = store.data.firstOrNull()
            themeOverride = parse(prefs?.get(themeKey))
            val raw = prefs?.get(modelKey).orEmpty()
            defaultModelID = if (raw.isNotEmpty() && ModelCatalog.find(raw) != null) {
                raw
            } else {
                ModelCatalog.fallback.id
            }
        }
    }

    fun setTheme(value: ThemeOverride) {
        themeOverride = value
        viewModelScope.launch {
            store.edit { prefs -> prefs[themeKey] = value.name }
        }
    }

    fun setDefaultModel(id: String) {
        defaultModelID = id
        viewModelScope.launch {
            store.edit { prefs -> prefs[modelKey] = id }
        }
    }

    private fun parse(raw: String?): ThemeOverride =
        when (raw) {
            ThemeOverride.LIGHT.name -> ThemeOverride.LIGHT
            ThemeOverride.DARK.name  -> ThemeOverride.DARK
            else                     -> ThemeOverride.SYSTEM
        }
}
